Context.js

```diff
@@ -0,0 +1,219 @@
+'use strict';
+
+const EventEmitter = require('events');
+const IncomingWebSocket = require('./IncomingWebSocket');
+const OutgoingWebSocket = require('./OutgoingWebSocket');
+const logger = require('./loggerFactory');
+const { createTarget } = require('./util');
+const {
+  kSender,
+  kUpstreamClosed,
+  kStartConnection,
+  kMessageReceived,
+  kError,
+  kSendMessage,
+  kQueueMessage,
+  kDrainMessage,
+  kConnectionOpened,
+  kDequeueMessage,
+  kClientClosed,
+  kCleanup,
+  kDrainCompleted,
+  kReleaseTap,
+  kAddNewContext,
+  kUpstreamRestart,
+  config,
+  INCOMING,
+  OUTGOING
+} = require('./constants');
+
+/**
+ * Context holds the information for the incoming and outgoing sockets.
+ *
+ * Instantiates the incoming and outgoing sockets and registers the listeners.
+ * Acts as a bridge between incoming and outgoing sockets.
+ */
+class Context extends EventEmitter {
+
+  /**
+   * Creates the context with connection identifier.
+   *
+   * @param {string} connectionId
+   */
+  constructor(connectionId) {
+    super();
+    this.connectionId = connectionId;
+    this.incomingSocket = null;
+    this.outgoingSocket = null;
+    this.incomingLock = true;
+    this.outgoingLock = false;
+  }
+
+  /**
+   * Adds a new incoming and outgoing connection.
+   *
+   * @param {WebSocket} socket
+   * @param {object} request
+   */
+  addNewConnection(socket, request) {
+    const prevIncomingSocket = this.incomingSocket === null;
+    if (prevIncomingSocket) {
+      this.incomingSocket = new IncomingWebSocket(socket, request);
+      this.outgoingSocket = new OutgoingWebSocket(createTarget(request.url), request.headers);
+      this.outgoingSocket.addSocket();
+      this.registerIncomingListeners();
+      this.registerOutgoingListeners();
+      this.incomingSocket.emit(kStartConnection);
+    } else {
+      this.incomingSocket.setSocket(socket, request);
+      this.outgoingSocket.emit(kDequeueMessage);
+      clearTimeout(this.upstreamCloseTimer);
+      this.upstreamCloseTimer = null;
+    }
+  }
+
+  /**
+   * Registers the incoming socket listeners.
+   */
+  registerIncomingListeners() {
+    this.incomingSocket.on(kMessageReceived, msg => {
+      if (this.incomingLock) {
+        logger.debug(`${INCOMING} [${this.connectionId}] [QUEUE] - ${msg}`);
+        this.incomingSocket.addToQueue(msg);
+      }
+      else
```

Comment:
Let's add a basic linter to the code and run `--fix` once.

Reply:
added